gluten-substrait/src/main/scala/org/apache/gluten/planner/metadata/OutputPartitioning.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.planner.metadata
+
+import org.apache.gluten.planner.metadata.GlutenMetadataModel.Verifier
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.plans.physical.Partitioning
+
+case class OutputPartitioning(partitioning: Partitioning) {
+  override def hashCode(): Int = System.identityHashCode(partitioning)
+  override def equals(obj: Any): Boolean = obj match {
+    case OutputPartitioning(otherPartitioning) => partitioning.equals(otherPartitioning)
+    case _ => false
+  }
+
+  override def toString: String = s"$partitioning"
+}
```

Comment:
Speaking of query optimizer, ideally `OutputPartitioning` should be categorized into a physical [`Property`](https://github.com/apache/incubator-gluten/blob/main/gluten-ras/common/src/main/scala/org/apache/gluten/ras/PropertyModel.scala) than into [`Metadata`](https://github.com/apache/incubator-gluten/blob/main/gluten-ras/common/src/main/scala/org/apache/gluten/ras/MetadataModel.scala).

But the implementation can be trickier. Is this inevitable to add this code?

Reply:
> Is this inevitable to add 'OutputPartitioning' for implementing this feature?

I think so. We need the partitioning information to decide whether a shuffle is needed. Otherwise, we can only add the shuffle blindly and then add another rule to remove unnecessary ones afterwards somewhere else.

Reply:
Thank you for the explanation. And I'll do some local tests with the change until this weekend then get back to you.

Reply:
@ivoson There was an old attempt from my side for incorporating `OutputPartitioning` / `OutputOrdering` into RAS planner.

[1] https://github.com/zhztheplayer/gluten/commit/d1575cce3626eaf4efe8c06924d108008e6e2c1e#diff-305f3a48d262389cc311d17f15cd27f6a1efc0d6da2dcd0ebe75975ccf2e9447 

This PR's solution is OK to me as of now, but I think we can do a refactor in future to use properties rather than metadata.

Or I think we can keep the PR open and take some time to look back on `[1]` to see if there are still blockers against the effort. I feel we should make sure the refactor will be viable then it's safe to merge the current PR's code as a temporary solution.

Reply:
> Or I think we can keep the PR open and take some time to look back on [1] to see if there are still blockers against the effort.

Hi @zhztheplayer , do you mean any blockers to incorporate  OutputPartitioning / OutputOrdering into RAS planner? Anything I can help here?